File: tools/tf-mp2-editor-csharp/TfMp2Editor/DbService.cs
using Microsoft.Data.SqlClient;

namespace TfMp2Editor;

internal sealed class ViewMetadata
{
    public List<string> Columns { get; init; } = [];
    public string MpNoColumn { get; init; } = "";
    public string SoNoColumn { get; init; } = "";
    public string KeyMpColumn { get; init; } = "";
    public string KeyItmColumn { get; init; } = "";
    public string ChkColumn { get; init; } = "";
}

internal sealed class GridRowModel
{
    public Dictionary<string, object?> Raw { get; init; } = new(StringComparer.OrdinalIgnoreCase);
    public bool Checked { get; set; }
    public bool OrigChecked { get; set; }
    public bool DeleteMark { get; set; }
    public string OrigPrdNoChg { get; set; } = "";

    public object? MpNo => Raw.TryGetValue(KeyMp, out var v) ? v : null;
    public object? Itm => Raw.TryGetValue(KeyItm, out var v) ? v : null;

    internal string KeyMp { get; set; } = AppConstants.KeyMpNo;
    internal string KeyItm { get; set; } = AppConstants.KeyItm;

    public bool CheckboxDirty => Checked != OrigChecked;

    public string DisplayValue(string column)
    {
        if (!Raw.TryGetValue(column, out var val) || val is null || val is DBNull)
            return "";
        if (val is DateTime dt) return dt.ToString("yyyy-MM-dd HH:mm:ss");
        return Convert.ToString(val) ?? "";
    }
}

internal sealed class DbService
{
    private readonly DbConfig _cfg;
    public ViewMetadata Meta { get; private set; } = new();

    public DbService(DbConfig cfg)
    {
        _cfg = cfg;
    }

    private SqlConnection Open() => new(_cfg.ConnectionString);

    private static string? PickColumn(IReadOnlyList<string> columns, IEnumerable<string> candidates)
    {
        var map = columns.ToDictionary(c => c, c => c, StringComparer.OrdinalIgnoreCase);
        foreach (var name in candidates)
            if (map.TryGetValue(name, out var actual))
                return actual;
        return null;
    }

    public void TestConnection()
    {
        using var conn = Open();
        conn.Open();
    }

    public void InitMetadata()
    {
        using var conn = Open();
        conn.Open();
        var columns = LoadViewColumns(conn);
        var mp = PickColumn(columns, AppConstants.MpNoCandidates)
            ?? throw new InvalidOperationException("视图中未找到「分析单号」或 MP_NO 字段");
        var so = PickColumn(columns, AppConstants.SoNoCandidates)
            ?? throw new InvalidOperationException("视图中未找到「订单号」或 SO_NO 字段");
        var keyMp = PickColumn(columns, new[] { AppConstants.KeyMpNo, "mp_no" })
            ?? throw new InvalidOperationException("视图中未找到 MP_NO 关联字段");
        var keyItm = PickColumn(columns, new[] { AppConstants.KeyItm, "itm" })
            ?? throw new InvalidOperationException("视图中未找到 ITM 关联字段");
        var chk = PickColumn(columns, new[] { AppConstants.ChkField, "prd_no_chg" })
            ?? throw new InvalidOperationException("视图中未找到 PRD_NO_CHG 字段");

        Meta = new ViewMetadata
        {
            Columns = columns,
            MpNoColumn = mp,
            SoNoColumn = so,
            KeyMpColumn = keyMp,
            KeyItmColumn = keyItm,
            ChkColumn = chk,
        };
    }

    private static List<string> LoadViewColumns(SqlConnection conn)
    {
        var cols = new List<string>();
        using (var cmd = conn.CreateCommand())
        {
            cmd.CommandText = """
                SELECT c.COLUMN_NAME
                FROM INFORMATION_SCHEMA.COLUMNS c
                WHERE c.TABLE_NAME = @table
                ORDER BY c.ORDINAL_POSITION
                """;
            cmd.Parameters.AddWithValue("@table", AppConstants.ViewName);
            using var reader = cmd.ExecuteReader();
            while (reader.Read())
                cols.Add(reader.GetString(0));
        }

        if (cols.Count == 0)
        {
            using var cmd = conn.CreateCommand();
            cmd.CommandText = $"SELECT TOP 0 * FROM {AppConstants.ViewName}";
            using var reader = cmd.ExecuteReader();
            for (var i = 0; i < reader.FieldCount; i++)
                cols.Add(reader.GetName(i));
        }

        if (cols.Count == 0)
            throw new InvalidOperationException($"视图 {AppConstants.ViewName} 不存在或无法读取列信息");

        return cols;
    }

    public List<string> FetchDistinctMpNos()
    {
        return FetchDistinct(Meta.MpNoColumn);
    }

    public List<string> FetchDistinctSoNos()
    {
        return FetchDistinct(Meta.SoNoColumn);
    }

    private List<string> FetchDistinct(string column)
    {
        using var conn = Open();
        conn.Open();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = $"""
            SELECT DISTINCT CAST([{column}] AS nvarchar(100)) AS v
            FROM {AppConstants.ViewName}
            WHERE [{column}] IS NOT NULL AND CAST([{column}] AS nvarchar(100)) <> ''
            ORDER BY v
            """;
        var list = new List<string>();
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
        {
            var v = reader.IsDBNull(0) ? "" : reader.GetString(0).Trim();
            if (!string.IsNullOrEmpty(v)) list.Add(v);
        }
        return list;
    }

    public List<GridRowModel> QueryRows(string? mpNo, string? soNo)
    {
        using var conn = Open();
        conn.Open();
        using var cmd = conn.CreateCommand();
        var where = new List<string>();
        if (!string.IsNullOrWhiteSpace(mpNo))
        {
            where.Add($"CAST([{Meta.MpNoColumn}] AS nvarchar(100)) = @mp_no");
            cmd.Parameters.AddWithValue("@mp_no", mpNo.Trim());
        }
        if (!string.IsNullOrWhiteSpace(soNo))
        {
            where.Add($"CAST([{Meta.SoNoColumn}] AS nvarchar(100)) = @so_no");
            cmd.Parameters.AddWithValue("@so_no", soNo.Trim());
        }

        cmd.CommandText = $"SELECT * FROM {AppConstants.ViewName}";
        if (where.Count > 0)
            cmd.CommandText += " WHERE " + string.Join(" AND ", where);
        cmd.CommandText += $" ORDER BY [{Meta.KeyMpColumn}], [{Meta.KeyItmColumn}]";

        var rows = new List<GridRowModel>();
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
        {
            var raw = new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase);
            for (var i = 0; i < reader.FieldCount; i++)
                raw[reader.GetName(i)] = reader.IsDBNull(i) ? null : reader.GetValue(i);

            var chkVal = raw.TryGetValue(Meta.ChkColumn, out var c) ? c : null;
            var checkedState = IsChecked(chkVal);
            rows.Add(new GridRowModel
            {
                Raw = raw,
                Checked = checkedState,
                OrigChecked = checkedState,
                OrigPrdNoChg = FormatVal(chkVal),
                KeyMp = Meta.KeyMpColumn,
                KeyItm = Meta.KeyItmColumn,
            });
        }
        return rows;
    }

    public int DeleteRows(IEnumerable<(object? MpNo, object? Itm)> keys)
    {
        var keyList = keys.ToList();
        if (keyList.Count == 0) return 0;

        using var conn = Open();
        conn.Open();
        using var tx = conn.BeginTransaction();
        var deleted = 0;
        try
        {
            foreach (var (mpNo, itm) in keyList)
            {
                using var cmd = conn.CreateCommand();
                cmd.Transaction = tx;
                cmd.CommandText = $"""
                    DELETE FROM {AppConstants.TableName}
                    WHERE [{AppConstants.KeyMpNo}] = @mp_no AND [{AppConstants.KeyItm}] = @itm
                    """;
                cmd.Parameters.AddWithValue("@mp_no", mpNo ?? DBNull.Value);
                cmd.Parameters.AddWithValue("@itm", itm ?? DBNull.Value);
                deleted += cmd.ExecuteNonQuery();
            }
            tx.Commit();
            return deleted;
        }
        catch
        {
            tx.Rollback();
            throw;
        }
    }

    public void UpdatePrdNoChg(object? mpNo, object? itm, string value)
    {
        using var conn = Open();
        conn.Open();
        using var cmd = conn.CreateCommand();
        cmd.CommandText = $"""
            UPDATE {AppConstants.TableName}
            SET [{AppConstants.ChkField}] = @val
            WHERE [{AppConstants.KeyMpNo}] = @mp_no AND [{AppConstants.KeyItm}] = @itm
            """;
        cmd.Parameters.AddWithValue("@val", value);
        cmd.Parameters.AddWithValue("@mp_no", mpNo ?? DBNull.Value);
        cmd.Parameters.AddWithValue("@itm", itm ?? DBNull.Value);
        cmd.ExecuteNonQuery();
    }

    public static bool IsChecked(object? val)
    {
        if (val is null or DBNull) return false;
        return !string.IsNullOrWhiteSpace(Convert.ToString(val));
    }

    public static string CheckedValue(object? original, bool wantChecked)
    {
        if (!wantChecked) return "";
        var s = original is null or DBNull ? "" : Convert.ToString(original)?.Trim() ?? "";
        return string.IsNullOrEmpty(s) ? "Y" : s;
    }

    private static string FormatVal(object? val)
    {
        if (val is null or DBNull) return "";
        return Convert.ToString(val) ?? "";
    }
}

File: tools/tf-mp2-editor/db_service.py
"""MSSQL 访问：VIEW_TF_MP2 查询与 TF_MP2 更新/删除。"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import pyodbc

from config import (
    CHK_FIELD,
    KEY_ITM,
    KEY_MP_NO,
    MP_NO_CANDIDATES,
    SO_NO_CANDIDATES,
    TABLE_NAME,
    VIEW_NAME,
)


@dataclass
class DbConfig:
    server: str
    port: str
    database: str
    user: str
    password: str

    def connection_string(self) -> str:
        host = self.server.strip()
        port = str(self.port or '').strip()
        if port and ',' not in host and '\\' not in host:
            host = f'{host},{port}'
        return (
            'DRIVER={ODBC Driver 17 for SQL Server};'
            f'SERVER={host};'
            f'DATABASE={self.database};'
            f'UID={self.user};'
            f'PWD={self.password};'
            'TrustServerCertificate=yes;'
        )


class DatabaseError(Exception):
    pass


def _pick_column(columns: list[str], candidates: tuple[str, ...]) -> str | None:
    lower_map = {c.lower(): c for c in columns}
    for name in candidates:
        if name.lower() in lower_map:
            return lower_map[name.lower()]
    return None


class DbService:
    def __init__(self, cfg: DbConfig):
        self.cfg = cfg
        self._columns: list[str] = []
        self.mp_no_col: str | None = None
        self.so_no_col: str | None = None
        self.key_mp_col: str | None = None
        self.key_itm_col: str | None = None
        self.chk_col: str | None = None

    def connect(self) -> pyodbc.Connection:
        drivers = [
            'ODBC Driver 18 for SQL Server',
            'ODBC Driver 17 for SQL Server',
            'ODBC Driver 13 for SQL Server',
            'SQL Server',
        ]
        last_err: Exception | None = None
        base = self.cfg.connection_string()
        for drv in drivers:
            try:
                conn_str = base.replace('ODBC Driver 17 for SQL Server', drv)
                return pyodbc.connect(conn_str, timeout=8)
            except Exception as e:
                last_err = e
        raise DatabaseError(f'无法连接数据库：{last_err}')

    def test_connection(self) -> None:
        conn = self.connect()
        conn.close()

    def _load_view_columns(self, conn: pyodbc.Connection) -> list[str]:
        cur = conn.cursor()
        cur.execute(
            """
            SELECT c.COLUMN_NAME
            FROM INFORMATION_SCHEMA.COLUMNS c
            WHERE c.TABLE_NAME = ?
            ORDER BY c.ORDINAL_POSITION
            """,
            VIEW_NAME,
        )
        cols = [row[0] for row in cur.fetchall()]
        if not cols:
            cur.execute(f'SELECT TOP 0 * FROM {VIEW_NAME}')
            cols = [d[0] for d in cur.description]
        if not cols:
            raise DatabaseError(f'视图 {VIEW_NAME} 不存在或无法读取列信息')
        return cols

    def init_metadata(self) -> None:
        conn = self.connect()
        try:
            self._columns = self._load_view_columns(conn)
            self.mp_no_col = _pick_column(self._columns, MP_NO_CANDIDATES)
            self.so_no_col = _pick_column(self._columns, SO_NO_CANDIDATES)
            self.key_mp_col = _pick_column(self._columns, (KEY_MP_NO, 'mp_no'))
            self.key_itm_col = _pick_column(self._columns, (KEY_ITM, 'itm'))
            self.chk_col = _pick_column(self._columns, (CHK_FIELD, 'prd_no_chg'))
            if not self.mp_no_col:
                raise DatabaseError('视图中未找到「分析单号」或 MP_NO 字段')
            if not self.so_no_col:
                raise DatabaseError('视图中未找到「订单号」或 SO_NO 字段')
            if not self.key_mp_col or not self.key_itm_col:
                raise DatabaseError('视图中未找到 MP_NO / ITM 关联字段，无法回写 TF_MP2')
            if not self.chk_col:
                raise DatabaseError('视图中未找到 PRD_NO_CHG 字段')
        finally:
            conn.close()

    @property
    def display_columns(self) -> list[str]:
        """表格展示列：用「选择」代替 PRD_NO_CHG 原列。"""
        out: list[str] = []
        for col in self._columns:
            if col == self.chk_col:
                out.append('__CHK__')
            else:
                out.append(col)
        return out

    def column_label(self, col: str) -> str:
        if col == '__CHK__':
            return '选择'
        return col

    def fetch_distinct_mp_nos(self) -> list[str]:
        conn = self.connect()
        try:
            cur = conn.cursor()
            cur.execute(
                f"""
                SELECT DISTINCT CAST([{self.mp_no_col}] AS nvarchar(100)) AS v
                FROM {VIEW_NAME}
                WHERE [{self.mp_no_col}] IS NOT NULL AND CAST([{self.mp_no_col}] AS nvarchar(100)) <> ''
                ORDER BY v
                """
            )
            return [str(r[0]).strip() for r in cur.fetchall() if r[0] is not None]
        finally:
            conn.close()

    def fetch_distinct_so_nos(self) -> list[str]:
        conn = self.connect()
        try:
            cur = conn.cursor()
            cur.execute(
                f"""
                SELECT DISTINCT CAST([{self.so_no_col}] AS nvarchar(100)) AS v
                FROM {VIEW_NAME}
                WHERE [{self.so_no_col}] IS NOT NULL AND CAST([{self.so_no_col}] AS nvarchar(100)) <> ''
                ORDER BY v
                """
            )
            return [str(r[0]).strip() for r in cur.fetchall() if r[0] is not None]
        finally:
            conn.close()

    def query_rows(self, mp_no: str | None, so_no: str | None) -> list[dict[str, Any]]:
        conn = self.connect()
        try:
            where: list[str] = []
            params: list[Any] = []
            if mp_no:
                where.append(f'CAST([{self.mp_no_col}] AS nvarchar(100)) = ?')
                params.append(mp_no)
            if so_no:
                where.append(f'CAST([{self.so_no_col}] AS nvarchar(100)) = ?')
                params.append(so_no)
            sql = f'SELECT * FROM {VIEW_NAME}'
            if where:
                sql += ' WHERE ' + ' AND '.join(where)
            sql += f' ORDER BY [{self.key_mp_col}], [{self.key_itm_col}]'
            cur = conn.cursor()
            cur.execute(sql, params)
            rows: list[dict[str, Any]] = []
            for rec in cur.fetchall():
                item = {col: rec[idx] for idx, col in enumerate(self._columns)}
                rows.append(item)
            return rows
        finally:
            conn.close()

    def delete_rows(self, keys: list[tuple[Any, Any]]) -> int:
        if not keys:
            return 0
        conn = self.connect()
        try:
            cur = conn.cursor()
            deleted = 0
            for mp_no, itm in keys:
                cur.execute(
                    f'DELETE FROM {TABLE_NAME} WHERE [{KEY_MP_NO}] = ? AND [{KEY_ITM}] = ?',
                    mp_no,
                    itm,
                )
                deleted += cur.rowcount
            conn.commit()
            return deleted
        except Exception:
            conn.rollback()
            raise
        finally:
            conn.close()

    def update_prd_no_chg(self, mp_no: Any, itm: Any, value: str) -> None:
        conn = self.connect()
        try:
            cur = conn.cursor()
            cur.execute(
                f'UPDATE {TABLE_NAME} SET [{CHK_FIELD}] = ? WHERE [{KEY_MP_NO}] = ? AND [{KEY_ITM}] = ?',
                value,
                mp_no,
                itm,
            )
            conn.commit()
        except Exception:
            conn.rollback()
            raise
        finally:
            conn.close()

    @staticmethod
    def is_checked(val: Any) -> bool:
        if val is None:
            return False
        return str(val).strip() != ''

    @staticmethod
    def checked_value(original: Any, want_checked: bool) -> str:
        if want_checked:
            if original is not None and str(original).strip():
                return str(original).strip()
            return 'Y'
        return ''
